# Lab 2: Iris

## Overview

Iris Flowers Classification with Logistical Regresion.

Team
- Victor Manuel Castrejon Morales
- Hiram Javier Torres Lopez

## Training/Test

The first step is to download the dataset from the sklearn collection.

Then we are going to split the dataset into an `X` matrix with atributes, and an array `y` with labels with the following distribution:

Training    | Test  |
:-----------|-------|
30          | 70    |
50          | 50    |
70          | 30    |

In [39]:
# Initial setup
import sklearn
assert sklearn.__version__ >= "0.20"

# Common imports
import numpy as np
import os

# to make this notebook's output stable across runs
np.random.seed(42)

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

iris = load_iris()
X = iris.data
y = iris.target

### 30/70

In [40]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.70, random_state=42)

In [41]:
from sklearn.linear_model import LogisticRegression

log_reg = LogisticRegression(multi_class="multinomial", solver="lbfgs", random_state=42, penalty="none")
log_reg.fit(X_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [42]:
y_proba = log_reg.predict_proba(X_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.9619047619047619

### 50/50

In [43]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.50, random_state=42)

In [44]:
log_reg.fit(X_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [45]:
y_proba = log_reg.predict_proba(X_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.96

### 70/30

In [46]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)

In [47]:
log_reg.fit(X_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [48]:
y_proba = log_reg.predict_proba(X_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

1.0

Training    | Test  | Accuracy score    |
:-----------|-------| ------------------|
30          | 70    | 0.9619            |
50          | 50    | 0.91              |
70          | 30    | 1.0               |

From those results, we can confirm the best way to split our dataset for LogisticRegression, **without any penalization**, is 70 training and 30 test.

## Single feature classification

In [49]:
iris.feature_names


['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

**sepal length (cm)**

In [54]:
x_sepal_length = X[:, 0]

In [57]:
x_sepal_length.size

150

In [74]:
X_train, X_test, y_train, y_test = train_test_split(x_sepal_length, y, test_size=0.30, random_state=42)

# Reshape your data either using array.reshape(-1, 1) if your data has a single feature.
X_train = X_train.reshape(-1, 1)
X_test = X_test.reshape(-1, 1)

In [77]:
log_reg = LogisticRegression(multi_class="multinomial", solver="lbfgs", random_state=42, penalty="none")
log_reg.fit(X_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [78]:
y_proba = log_reg.predict_proba(X_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.6888888888888889

**sepal width (cm)**

In [79]:
x_sepal_width = X[:, 1]

In [80]:
X_train, X_test, y_train, y_test = train_test_split(x_sepal_width, y, test_size=0.30, random_state=42)

# Reshape your data either using array.reshape(-1, 1) if your data has a single feature.
X_train = X_train.reshape(-1, 1)
X_test = X_test.reshape(-1, 1)

In [81]:
log_reg.fit(X_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [82]:
y_proba = log_reg.predict_proba(X_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.5555555555555556

**petal length (cm)**

In [84]:
x_petal_length = X[:, 2]

In [85]:
X_train, X_test, y_train, y_test = train_test_split(x_petal_length, y, test_size=0.30, random_state=42)

# Reshape your data either using array.reshape(-1, 1) if your data has a single feature.
X_train = X_train.reshape(-1, 1)
X_test = X_test.reshape(-1, 1)

In [86]:
log_reg.fit(X_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [87]:
y_proba = log_reg.predict_proba(X_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

1.0

**petal width (cm)**

In [90]:
x_petal_width = X[:, 3]

In [92]:
X_train, X_test, y_train, y_test = train_test_split(x_petal_width, y, test_size=0.30, random_state=42)

# Reshape your data either using array.reshape(-1, 1) if your data has a single feature.
X_train = X_train.reshape(-1, 1)
X_test = X_test.reshape(-1, 1)

In [93]:
log_reg.fit(X_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [94]:
y_proba = log_reg.predict_proba(X_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

1.0

Feature             | Accuracy score    |
:-------------------| ------------------|
sepal length (cm)   | 0.6888            |
sepal width (cm)    | 0.5555            |
petal length (cm)   | 1.0               |
petal width (cm)    | 1.0               |

From those results, we can confirm petal length and width are the best features for LogisticRegression, **without any penalization**.

## Double feature classification

In [98]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)

**sepal length (cm), sepal width (cm)**

In [100]:
X_filtered_train = X_train[:, [0, 1]]
X_filtered_test = X_test[:, [0, 1]]

In [102]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [104]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.8222222222222222

**sepal length (cm), petal length (cm)**

In [107]:
X_filtered_train = X_train[:, [0, 2]]
X_filtered_test = X_test[:, [0, 2]]

In [108]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [109]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.9777777777777777

**sepal length (cm), petal width (cm)**

In [111]:
X_filtered_train = X_train[:, [0, 3]]
X_filtered_test = X_test[:, [0, 3]]

In [112]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [114]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

1.0

**sepal width (cm), petal length (cm)**

In [115]:
X_filtered_train = X_train[:, [1, 2]]
X_filtered_test = X_test[:, [1, 2]]

In [116]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [118]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.9777777777777777

**sepal width (cm), petal width (cm)**

In [119]:
X_filtered_train = X_train[:, [1, 3]]
X_filtered_test = X_test[:, [1, 3]]

In [121]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [122]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.9777777777777777

**petal length (cm), petal width (cm)**

In [123]:
X_filtered_train = X_train[:, [2, 3]]
X_filtered_test = X_test[:, [2, 3]]

In [124]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [125]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

1.0

Feature             | sepal length (cm) | sepal width (cm)  | petal length (cm) | petal width (cm)  |
:-------------------| ------------------| ------------------| ------------------| ------------------|
sepal length (cm)   | x                 | 0.8222            | 0.9777            | 1                 |
sepal width (cm)    | x                 | x                 | 0.9777            | 0.9777            |
petal length (cm)   | x                 | x                 | x                 | 1                 |
petal width (cm)    | x                 | x                 | x                 | x                 |

From those results, we can confirm petal width is the most successful feature, in combination with sepal and petal length, for LogisticRegression, **without any penalization**.

## Triple feature classification

**sepal length (cm), sepal width (cm), petal length (cm)**

In [126]:
X_filtered_train = X_train[:, [0, 1, 2]]
X_filtered_test = X_test[:, [0, 1, 2]]

In [127]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [128]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.9777777777777777

**sepal length (cm), sepal width (cm), petal width (cm)**

In [129]:
X_filtered_train = X_train[:, [0, 1, 3]]
X_filtered_test = X_test[:, [0, 1, 3]]

In [130]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [131]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.9777777777777777

**sepal width (cm), petal length (cm), petal width (cm)**

In [132]:
X_filtered_train = X_train[:, [1, 2, 3]]
X_filtered_test = X_test[:, [1, 2, 3]]

In [133]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [134]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

1.0

**petal length (cm), petal width (cm), sepal length (cm)**

In [135]:
X_filtered_train = X_train[:, [2, 3, 0]]
X_filtered_test = X_test[:, [2, 3, 0]]

In [136]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [137]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

1.0

**sepal length (cm), petal length (cm), petal width (cm)**

In [138]:
X_filtered_train = X_train[:, [1, 2, 3]]
X_filtered_test = X_test[:, [1, 2, 3]]

In [139]:
log_reg.fit(X_filtered_train, y_train)

LogisticRegression(multi_class='multinomial', penalty='none', random_state=42)

In [140]:
y_proba = log_reg.predict_proba(X_filtered_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

1.0

From those results, we can't confirm which triple combination works better. We can notice that petal length and width combined are better.

## Penalty 

### L1 

In [142]:
log_reg = LogisticRegression(multi_class="ovr", solver="liblinear", random_state=42, penalty="l1")
log_reg.fit(X_train, y_train)

LogisticRegression(multi_class='ovr', penalty='l1', random_state=42,
                   solver='liblinear')

In [143]:
y_proba = log_reg.predict_proba(X_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

1.0

### L2

In [145]:
log_reg = LogisticRegression(multi_class="ovr", solver="liblinear", random_state=42, penalty="l2")
log_reg.fit(X_train, y_train)

LogisticRegression(multi_class='ovr', random_state=42, solver='liblinear')

In [146]:
y_proba = log_reg.predict_proba(X_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.9777777777777777

### Elasticnet

In [160]:
log_reg = LogisticRegression(multi_class="ovr", solver="saga", random_state=42, penalty="elasticnet", l1_ratio=0.01, max_iter=4000)
log_reg.fit(X_train, y_train)

LogisticRegression(l1_ratio=0.01, max_iter=4000, multi_class='ovr',
                   penalty='elasticnet', random_state=42, solver='saga')

In [161]:
y_proba = log_reg.predict_proba(X_test)
y_predict = np.argmax(y_proba, axis=1)

accuracy_score = np.mean(y_predict == y_test)
accuracy_score

0.9555555555555556

L1  | L2        | Elasticnet    | None  |
:---|-----------| --------------|-------|
1.0 | 0.9777    | 0.9555        | 1.0   |

From those results, we can have good results without penalization. Also, L1 works just fine.